Notlar için gerekli olan düzenlemeler yapıldı

AbdullahOztuurkk opened the pull request:
-- Added Note Class
-- Added Note Repository
-- Added Note Validator
-- Added INoteDetails Interface
-- Added INoteRepository Interface
-- Added NoteCollection Class
-- Added NoteController

AbdullahOztuurkk committed:
-- Added NoteDto
-- Added Note Class
-- Added Note Repository
-- Added Note Validator
-- Added INoteDetails Interface
-- Added INoteRepository Interface
-- Added NoteCollection Class
-- Added NoteController

fasetto commented:
@AbdullahOztuurkk Koda henüz bakamadım ama, CI da build fail olmuş mesela. 
```
Microsoft (R) Build Engine version 16.6.0+5ff7b0c9e for .NET Core
Copyright (C) Microsoft Corporation. All rights reserved.

  DevNots.Domain -> /home/runner/work/devnots/devnots/src/DevNots.Domain/bin/Release/netstandard2.0/DevNots.Domain.dll
##[error]MongoDbContext.cs(6,18): error CS0534: 'MongoDbContext' does not implement inherited abstract member 'DbContext.Notes.get' [/home/runner/work/devnots/devnots/src/DevNots.MongoDb/DevNots.MongoDb.csproj]

Build FAILED.

##[error]MongoDbContext.cs(6,18): error CS0534: 'MongoDbContext' does not implement inherited abstract member 'DbContext.Notes.get' [/home/runner/work/devnots/devnots/src/DevNots.MongoDb/DevNots.MongoDb.csproj]
```
Abstract class daki implement edilmesi gereken bir şeyi unutmuş olabilirsin diyor hata mesajında.
Şimdi kodu bi inceleyeyim.

fasetto commented on src/DevNots.Application/Services/UserService.cs:
```diff
@@ -27,12 +27,12 @@ public UserService(IUserRepository userRepository, IMapper mapper, IValidator<Us
-                var errorMessage = validatitonResult.Errors.FirstOrDefault().ErrorMessage;
+                var errorMessage = validationResult.Errors.FirstOrDefault().ErrorMessage;
                 return ErrorResponse(errorMessage, 400, response);
             }
```
Bunu yakaladığın iyi olmuş. (:

fasetto commented on src/DevNots.Application/Validations/NoteValidator.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public NoteValidator()
+        {
+            RuleFor(rule => rule.Text).MinimumLength(1);
```
Burada minimum length yerine max length kullanırsak daha iyi olabilir. Ya da böyle de kalabilir bilemiyorum.

AbdullahOztuurkk committed:
MongoDb tanımlaması yapildi

fasetto commented on src/DevNots.Application/Contracts/NoteDto.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace DevNots.Application.Contracts
+{
+    public class NoteDto
```
INoteDetails interface ini implement etmeli burası

fasetto commented on src/DevNots.Application/Services/NoteService.cs:
```diff
@@ -0,0 +1,30 @@
+using DevNots.Domain.Note;
+using FluentValidation;
+
+namespace DevNots.Application.Services
```
Buradaki eksik olan methodları da eklemek istersen devam edebilirsin.

fasetto replied:
Bence bununu da tamamla, merge ettiğimiz zaman canlıda 🌐 her şey hazır halde olsun.

fasetto commented on src/DevNots.Domain/Note/Note.cs:
```diff
@@ -0,0 +1,20 @@
+        public string Text { get; set; }
+        public DateTime CreatedAt { get; set; }
+
+        public Note(int UserId,string Text,DateTime CreatedAt)
```
createdAt parametresi constructordan kaldırılmalı. Bu parametre otomatik olarak doldurulur kayıt yaparken Dto tarafında

AbdullahOztuurkk replied:
PaginateAsync yapılmayacak mı?

fasetto replied:
Yapılacak fakat, createdAt property burada set etmek yerine NotDto da varsayılan olarak DateTime.UtcNow şeklinde atayabiliriz. 
Manuel olarak girmemiz gerekmez her defasında. (:

fasetto commented on src/DevNots.MongoDb/NoteCollection.cs:
```diff
@@ -0,0 +1,75 @@
+        public async Task<string> CreateAsync(Note aggregate)
+        {
+            await collection.InsertOneAsync(aggregate).ConfigureAwait(false);
+            var _note = await FindOneAsync(filter => filter.UserId == aggregate.UserId || filter.Text == aggregate.Text);
```
42. satırı aslında silebilirsin, hiç gerek yok buna.
`return aggregate.Id` şeklinde id döndürülmeli.

fasetto reviewed (changes requested):
Bahsettiğim değişiklikleri de yaparsın müsait olduğun da tekrar. 👍

AbdullahOztuurkk committed:
-- NoteDto , INoteDetails interface'inden implemente edildi.
 -- Note sınıfından constructor fonksiyonundan CreatedAt parametresi silindi
 -- NoteCollection sınıfındaki kod karmaşası giderildi. Dönüş parametresi olarak Id gönderildi.

fasetto commented on src/DevNots.Domain/Note/INoteDetails.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public interface INoteDetails
+    {
+        public int UserId { get; set; }
```
UserId int değil de string tipinde olmalı. Çünki MongoDb de ObjectId tipinde tutuluyor aslında bu değer. Kod tarafında stringe dönüştürülüp o şekilde kullanıyoruz.

fasetto commented on src/DevNots.Domain/Note/Note.cs:
```diff
@@ -0,0 +1,19 @@
+{
+    public class Note:AggregateRoot,INoteDetails
+    {
+        public int UserId { get; set; }
```
Burası da aynı şekilde string tipinde olmalı.

fasetto commented:
Merge yapılmadan issue ları kapamıyoruz bu arada. Linklendiği zaman birbirine issue ve pull request, merge yapıldığı zaman otomatik olarak kapanıyor. 😅

AbdullahOztuurkk committed:
ID parametresinin türü değiştirildi.

AbdullahOztuurkk committed:
-- NoteDto'daki UserId parametresinin türü degistirildi.

fasetto commented on src/DevNots.Domain/Note/INoteDetails.cs:
```diff
@@ -0,0 +1,12 @@
+    public interface INoteDetails
+    {
+        public string UserId { get; set; }
+        public string Text { get; set; }
```
Burada olması gereken başka özellikler de var aslında. 
- [x] Title
- [x] Description gibi. 

Aklıma başka gelmedi ama eğer senin gelirse ekleyebilirsin (:
Bu arada interface de değişiklik yaptıktan sonra ilgili class ları düzenlemeyi unutma *(Note, NoteDto gibi)*

AbdullahOztuurkk committed:
-- Title , Description ve Keywords parametreleri eklendi.

fasetto commented:
@AbdullahOztuurkk Eline sağlık. Bitirdiysen merge yapabilirsin. 👍

AbdullahOztuurkk committed:
-- NoteService sınıfına yeni fonksiyonlar eklendi ve gerekli olan dto sınıfları yazıldı.
 -- NoteController sınıfı yazıldı.

fasetto reviewed (changes requested):
Her şey düzgün görünüyor. Ama bir şey var unuttuğun, şimdi contructorlarda tanımladığımız bazı bağımlılıklar var.
Bunları da inject etmek için IoC container kullanıyoruz. *(Microsoft.Extensions.DependencyInjection nuget paketi)*
Bu bağımlılıkları Application katmanında Extensions klasörü içinde ServiceCollectionExtensions var. Orada tanımlaman gerekiyor. UserService için tanımlamıştım, incelersen aynı şekilde kodları notlar için de yazmak gerekiyor. 😄
Yoksa runtime da hata verir bağımlılıkları çözemediği için. 
Kolay gelsin.

Edit: [ServiceCollectionExtensions.cs](https://github.com/fasetto/devnots/blob/master/src/DevNots.Application/Extensions/ServiceCollectionExtensions.cs)

AbdullahOztuurkk commented:
> Her şey düzgün görünüyor. Ama bir şey var unuttuğun, şimdi contructorlarda tanımladığımız bazı bağımlılıklar var.
> Bunları da inject etmek için IoC container kullanıyoruz. _(Microsoft.Extensions.DependencyInjection nuget paketi)_
> Bu bağımlılıkları Application katmanında Extensions klasörü içinde ServiceCollectionExtensions var. Orada tanımlaman gerekiyor. UserService için tanımlamıştım, incelersen aynı şekilde kodları notlar için de yazmak gerekiyor. 😄
> Yoksa runtime da hata verir bağımlılıkları çözemediği için.
> Kolay gelsin.
> 
> Edit: [ServiceCollectionExtensions.cs](https://github.com/fasetto/devnots/blob/master/src/DevNots.Application/Extensions/ServiceCollectionExtensions.cs)

Gözden kaçırmışım pardon ::)
Hemen düzenleyip gönderiyorum

AbdullahOztuurkk committed:
-- Dependency Injection yöntemi kullanıldı

fasetto reviewed (approved):
Eline sağlık, merge yapabilirsin. 👍
